### PR TITLE
crimson/os/seastore: reduce write amplification from record overhead and implement placement hint

### PR DESCRIPTION
**Pull request by cyx1231st**
This is a follow-up PR try to address the write gap observed from https://github.com/ceph/ceph/pull/41803#issuecomment-922243566.

This PR decides to inline allocate non-REWRITE extents in order to reduce the ool record overhead. The observation shows the transactions from user is usually very small and doesn't deserve to be written in ool records, which requires at least a block to store the record metadata.

The overall throughput is also improved in local tests.

## Results
1. This almost removes the write amplifaction from `OOL_RECORD_OVERHEAD` that takes ~0.7 * user-write-size:
Before:
before_1
After:
after_1
(not sure why the rate of `INLINE_RETIRED` is higher than before)

2. The overall gap between `SEGMENTED_WRITE` and `VALID_EXTENT_WRITE` is reduced by ~50%:
Before:
before_2
After:
after_2

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by athanatos** [comment]
make check has a build failure

File: src/crimson/os/seastore/extent_placement_manager.h [hunk 1]
```diff
@@ -409,7 +413,7 @@ class ExtentPlacementManager {
   }
 
   bool should_be_inline(LogicalCachedExtentRef& extent) {
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/extent_placement_manager.h]: done.

**Comment by athanatos** [comment]
@cyx1231st Probably worth adding some instrumentation to find how how full on average the record header block is.

**Comment by cyx1231st** [comment]
> Probably worth adding some instrumentation to find how how full on average the record header block is.

Yep, will do.

**Review by xxhdx1985126 — APPROVED** [review]

File: src/crimson/os/seastore/extent_placement_manager.h [hunk 2]
```diff
@@ -326,12 +328,14 @@ class ExtentPlacementManager {
...
     segment_off_t length,
-    placement_hint_t hint = placement_hint_t::NONE) {
+    placement_hint_t hint) {
     // only logical extents should fall in this path
     static_assert(is_logical_type(T::TYPE));
     assert(hint < placement_hint_t::NUM_HINTS);
     auto dtype = get_allocator_type(hint);
-    bool delay = can_delay_allocation(dtype);
+    // FIXME: set delay for COLD extent when the record overhead is low
+    bool delay = (hint > placement_hint_t::COLD &&
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/extent_placement_manager.h:337]: could you please check the comments and the condition check?  For COLD extent, hint will be  placement_hint_t::COLD, right? then delay is false here.

File: src/crimson/os/seastore/extent_placement_manager.h [hunk 3]
```diff
@@ -351,9 +355,8 @@ class ExtentPlacementManager {
     LOG_PREFIX(ExtentPlacementManager::delayed_alloc_or_ool_write);
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/extent_placement_manager.h:355]: please modify this function's description, lba update is removed. and this function just do delayed alloc, no ool_write, right?  can we just rename it delayed_allocate?

Comment (athanatos) [reply src/crimson/os/seastore/extent_placement_manager.h:355]: I'd actually like to leave that as is.  Once we start trying to optimize for larger RGW writes, I think we'll actually want to do ool writes here for large write sizes.

File: src/crimson/os/seastore/extent_placement_manager.h (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/extent_placement_manager.h:337]: Refer to the reply below, we probably want to do ool writes for COLD extents with larger RGW writes.

**Pull request merged**